### ISSUE TITLE
Search does not redirect.

### ISSUE THREAD

**Issue by shimizukawa** [sphinx-doc/sphinx#1591]
Hello.

I have only reproduced this in the C++ domain. Basically, using the search function on anything with namespaces, templates, or anything other than plain-text (i.e. [a-zA-Z0-9]) results in no redirect.

I think it's better to show it via gifs. I've tested this with v1.2.3 and v1.3.0a (sorry for the large gifs)

[sphinx-v1.3-bug.gif](https://bitbucket.org/repo/7zgG/images/1450789625-sphinx-v1.3-bug.gif)

[sphinx-v1.2-bug.gif](https://bitbucket.org/repo/7zgG/images/1708444518-sphinx-v1.2-bug.gif)

It seems that the issue is that search takes the string literally rather than its 'mangled' form.

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/1591
- Originally reported by: [Rapptz](https://bitbucket.org/Rapptz)
- Originally created at: 2014-10-09T13:38:37.923

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-10-09 12:24:10+00:00_

Thanks for reporting!

**Comment by shimizukawa** [comment]
_From [Rapptz](https://bitbucket.org/Rapptz) on 2014-10-09 23:19:34+00:00_

I'm not sure what putting the milestone to a future version means, I imagined it meant that it'd be put away as non-important based on the list of bugs in said milestone but I considered this bug to be pretty annoying so I investigated the code to fix it. After a couple hours of navigating seeing how search etc was implemented it seems like the code to fix it is rather simple.

Basically in this [line](https://github.com/sphinx-doc/sphinx/blob/master/sphinx/domains/cpp.py#L1896), the only thing that needs changing is `refname, 1)` to `ast.get_id(), 1)` then everything should work as expected. I think the change is too small for a pull-request since it could be done manually. I'm not sure if this breaks anything but from the testing I did (including running the tests with `make test`) nothing seemed to fail.

Thanks for giving me the motivation to look into fixing this myself and for the enjoyable tool for documenting.
